File: CS/ReportHelper/ReportHelper.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Printing;
using System.Linq;
using DevExpress.XtraPrinting;
using DevExpress.XtraPrinting.Native;
using DevExpress.XtraReports.UI;

namespace dxSample
{
    public class ReportHelper
    {
        private const float AdditionalCellWidth = 0;

        private static Dictionary<XRControl, float> cellColumnWidthCollection;
        /// <summary>
        /// Indicates whether or not the specified Printing System command can be handled.
        /// </summary>
        /// <param name="command">A <see cref="T:DevExpress.XtraPrinting.PrintingSystemCommand" /> enumeration value.</param>
        /// <param name="printControl">An object implementing the <see cref="T:DevExpress.XtraPrinting.IPrintControl" /> interface (typically, this is a <see cref="T:DevExpress.XtraPrinting.Control.PrintControl" /> instance).</param>
        /// <returns>
        ///   true, if the command can be handled; otherwise, false.
        /// </returns>
        public virtual bool CanHandleCommand(PrintingSystemCommand command, IPrintControl printControl)
        {
            bool canBeHandled = command == PrintingSystemCommand.ExportXlsx || command == PrintingSystemCommand.ExportXls || command == PrintingSystemCommand.ExportCsv;
            return canBeHandled;
        }

        public DevExpress.XtraReports.UI.XtraReport CreateReportWithAutoCellWidth(DevExpress.XtraReports.UI.XtraReport xtraReport)
        {
            // I need this for a later excel-export to simulate endless paper (all report pages on one sheet)
            xtraReport.PaperKind = PaperKind.Custom;
            xtraReport.ReportUnit = ReportUnit.Pixels;
            xtraReport.PageWidth = 3000;
            xtraReport.RollPaper = true;

            cellColumnWidthCollection = new Dictionary<XRControl, float>();
            xtraReport.CreateDocument();
            FillColumnCellWidthCollection(xtraReport);
            ApplyNewWidthToReportCells(xtraReport);
            xtraReport.CreateDocument();

            return xtraReport;
        }

        private void ApplyNewWidthToReportCells(DevExpress.XtraReports.UI.XtraReport xtraReport)
        {
            IList<KeyValuePair<XRTable, List<XRTable>>> tables = FindTables(xtraReport);
            foreach (var table in tables)
            {
                float totalWidth = 0;
                foreach (XRTableCell dc in table.Key.Rows[0].Cells)
                {
                    float newWidth = cellColumnWidthCollection[dc];
                    foreach (var otherTable in table.Value)
                    {
                        newWidth = Math.Max(newWidth, cellColumnWidthCollection[otherTable.Rows[0].Cells[dc.Index]]);
                    }

                    newWidth += AdditionalCellWidth;
                    totalWidth += newWidth;
                }

                table.Key.WidthF = totalWidth;
                foreach (var otherTable in table.Value)
                {
                    otherTable.WidthF = totalWidth;
                }

                table.Key.BeginInit();
                foreach (var otherTable in table.Value)
                {
                    otherTable.BeginInit();
                }

                foreach (XRTableCell dc in table.Key.Rows[0].Cells)
                {
                    float newWidth = cellColumnWidthCollection[dc];
                    foreach (var otherTable in table.Value)
                    {
                        newWidth = Math.Max(newWidth, cellColumnWidthCollection[otherTable.Rows[0].Cells[dc.Index]]);
                    }

                    newWidth += AdditionalCellWidth;
                    dc.WidthF = newWidth;
                    foreach (var otherTable in table.Value)
                    {
                        otherTable.Rows[0].Cells[dc.Index].WidthF = newWidth;
                    }
                }

                table.Key.EndInit();
                foreach (var otherTable in table.Value)
                {
                    otherTable.EndInit();
                }
            }
        }

        private IList<KeyValuePair<XRTable, List<XRTable>>> FindTables(DevExpress.XtraReports.UI.XtraReport xtraReport)
        {
            var result = new List<KeyValuePair<XRTable, List<XRTable>>>();
            IEnumerable<XRTable> allTables = xtraReport.AllControls<XRTable>().Where(t => t.Visible);
            foreach (var table in allTables)
            {
                var subResult = new List<XRTable>();
                XRTableCellCollection dataCells = table.Rows[0].Cells;

                bool match = true;
                IEnumerable<XRTable> candidates = allTables.Where(t => t.Visible && t != table);
                foreach (var candidate in candidates)
                {
                    if (candidate.Rows[0].Cells.Count == dataCells.Count)
                    {
                        XRTableCellCollection cells = candidate.Rows[0].Cells;
                        for (int i = 0; i < cells.Count; i++)
                        {
                            match &= FloatComparer.Compare(dataCells[i].WidthF, candidate.Rows[0].Cells[i].WidthF, 5) == 0;
                        }

                        if (match)
                        {
                            subResult.Add(candidate);
                        }
                    }
                }

                result.Add(new KeyValuePair<XRTable, List<XRTable>>(table, subResult));
            }

            return result;
        }

        private void FillColumnCellWidthCollection(DevExpress.XtraReports.UI.XtraReport xtraReport)
        {
            foreach (PSPage page in xtraReport.Pages)
            {
                NestedBrickIterator iterator = new NestedBrickIterator(page.InnerBricks);
                while (iterator.MoveNext())
                {
                    if (iterator.CurrentBrick is VisualBrick && ((VisualBrick)iterator.CurrentBrick).BrickOwner is XRTableCell)
                    {
                        string text = ((VisualBrick)iterator.CurrentBrick).Text;
                        XRTableCell cell = ((VisualBrick)iterator.CurrentBrick).BrickOwner as XRTableCell;
                        float actualWidth = BrickGraphics.MeasureString(text, cell.GetEffectiveFont(), 0, (StringFormat)StringFormat.GenericTypographic.Clone(), GraphicsUnit.Document).Width;

                        switch (xtraReport.ReportUnit)
                        {
                            case ReportUnit.HundredthsOfAnInch:
                                actualWidth = GraphicsUnitConverter.Convert(actualWidth, GraphicsUnit.Document, GraphicsUnit.Inch) * 100;
                                break;
                            case ReportUnit.TenthsOfAMillimeter:
                                actualWidth = GraphicsUnitConverter.Convert(actualWidth, GraphicsUnit.Document, GraphicsUnit.Millimeter) * 10;
                                break;
                            case ReportUnit.Pixels:
                                actualWidth = GraphicsUnitConverter.Convert(actualWidth, GraphicsUnit.Document, GraphicsUnit.Pixel);
                                break;
                        }
                        actualWidth += 1; //the returned text metrics are slighly off regarding what it actually takes to render the text

                        if (!cellColumnWidthCollection.ContainsKey(cell))
                        {
                            cellColumnWidthCollection.Add(cell, actualWidth);
                        }
                        else
                        {
                            float value = cellColumnWidthCollection[cell];
                            if (actualWidth > value)
                            {
                                cellColumnWidthCollection[cell] = actualWidth;
                            }
                        }
                    }
                }
            }
        }
    }
}

File: VB/ReportHelper/ReportHelper.vb
Imports DevExpress.XtraPrinting
Imports DevExpress.XtraPrinting.Native
Imports DevExpress.XtraReports.UI
Imports System
Imports System.Collections.Generic
Imports DevExpress.Drawing.Printing
Imports System.Linq

Namespace dxSample

    Public Class ReportHelper

        Private Const additionalCellSpace As Single = 3

        Private Shared cellColumnWidthCollection As Dictionary(Of XRControl, Single) = New Dictionary(Of XRControl, Single)()

        Public Function CreateReportWithAutoCellWidth(ByVal report As XtraReport) As XtraReport
            InitializeReport(report)
            CreateDocument(report)
            FillColumnCellWidthCollection(report)
            ApplyNewWidthToReportCells(report)
            CreateDocument(report)
            Return report
        End Function

        Private Sub CreateDocument(ByVal report As XtraReport)
            report.CreateDocument()
        End Sub

        Private Sub InitializeReport(ByVal targetReport As XtraReport)
            targetReport.PaperKind = DXPaperKind.Custom
            targetReport.ReportUnit = ReportUnit.Pixels
            targetReport.PageWidth = 3000
            targetReport.RollPaper = True
        End Sub

        Private Sub ApplyNewWidthToReportCells(ByVal report As XtraReport)
            Dim tables As IList(Of KeyValuePair(Of XRTable, List(Of XRTable))) = FindTables(report)
            For Each table In tables
                Dim totalWidth As Single = 0
                For Each dc As XRTableCell In table.Key.Rows(0).Cells
                    Dim newWidth As Single = cellColumnWidthCollection(dc)
                    For Each otherTable In table.Value
                        newWidth = Math.Max(newWidth, cellColumnWidthCollection(otherTable.Rows(0).Cells(dc.Index)))
                    Next

                    newWidth += additionalCellSpace
                    totalWidth += newWidth
                Next

                table.Key.WidthF = totalWidth
                For Each otherTable In table.Value
                    otherTable.WidthF = totalWidth
                Next

                table.Key.BeginInit()
                For Each otherTable In table.Value
                    otherTable.BeginInit()
                Next

                For Each dc As XRTableCell In table.Key.Rows(0).Cells
                    Dim newWidth As Single = cellColumnWidthCollection(dc)
                    For Each otherTable In table.Value
                        newWidth = Math.Max(newWidth, cellColumnWidthCollection(otherTable.Rows(0).Cells(dc.Index)))
                    Next

                    newWidth += additionalCellSpace
                    dc.WidthF = newWidth
                    For Each otherTable In table.Value
                        otherTable.Rows(0).Cells(dc.Index).WidthF = newWidth
                    Next
                Next

                table.Key.EndInit()
                For Each otherTable In table.Value
                    otherTable.EndInit()
                Next
            Next
        End Sub

        Private Function FindTables(ByVal xtraReport As XtraReport) As IList(Of KeyValuePair(Of XRTable, List(Of XRTable)))
            Dim result = New List(Of KeyValuePair(Of XRTable, List(Of XRTable)))()
            Dim allTables As IEnumerable(Of XRTable) = xtraReport.AllControls(Of XRTable)().Where(Function(t) t.Visible)
            For Each table In allTables
                Dim subResult = New List(Of XRTable)()
                Dim dataCells As XRTableCellCollection = table.Rows(0).Cells
                Dim match As Boolean = True
                Dim candidates As IEnumerable(Of XRTable) = allTables.Where(Function(t) t.Visible AndAlso t IsNot table)
                For Each candidate In candidates
                    If candidate.Rows(0).Cells.Count = dataCells.Count Then
                        Dim cells As XRTableCellCollection = candidate.Rows(0).Cells
                        For i As Integer = 0 To cells.Count - 1
                            match = match And Compare(dataCells(i).WidthF, candidate.Rows(0).Cells(i).WidthF, 5) = 0
                        Next

                        If match Then
                            subResult.Add(candidate)
                        End If
                    End If
                Next

                result.Add(New KeyValuePair(Of XRTable, List(Of XRTable))(table, subResult))
            Next

            Return result
        End Function

        Private Sub FillColumnCellWidthCollection(ByVal currentReport As XtraReport)
            For Each page As PSPage In currentReport.Pages
                Dim iterator As NestedBrickIterator = New NestedBrickIterator(page.InnerBricks)
                While iterator.MoveNext()
                    If TypeOf iterator.CurrentBrick Is VisualBrick AndAlso TypeOf CType(iterator.CurrentBrick, VisualBrick).BrickOwner Is XRTableCell Then
                        Dim cell As XRTableCell = TryCast(CType(iterator.CurrentBrick, VisualBrick).BrickOwner, XRTableCell)
                        Dim currentCellText As String = CType(iterator.CurrentBrick, VisualBrick).Text
                        Dim bestCellWidthForProvidedText As Single = BestSizeEstimator.GetBoundsToFitText(currentCellText, CType(iterator.CurrentBrick, VisualBrick).Style, currentReport.ReportUnit).Width
                        If Not cellColumnWidthCollection.ContainsKey(cell) Then
                            cellColumnWidthCollection.Add(cell, bestCellWidthForProvidedText)
                        Else
                            Dim value As Single = cellColumnWidthCollection(cell)
                            If bestCellWidthForProvidedText > value Then
                                cellColumnWidthCollection(cell) = bestCellWidthForProvidedText
                            End If
                        End If
                    End If
                End While
            Next
        End Sub
    End Class
End Namespace
